core/src/main/java/hudson/model/FingerprintCleanupThread.java

```diff
@@ -66,13 +70,13 @@ private static FingerprintCleanupThread getInstance() {
     public void execute(TaskListener listener) {
         int numFiles = 0;
 
-        File root = new File(Jenkins.getInstance().getRootDir(),"fingerprints");
-        File[] files1 = root.listFiles(LENGTH2DIR_FILTER);
+        File root = new File(getRootDir(), FINGERPRINTS_DIR_NAME);
+        File[] files1 = root.listFiles(f -> f.isDirectory() && f.getName().length()==2);
         if(files1!=null) {
             for (File file1 : files1) {
-                File[] files2 = file1.listFiles(LENGTH2DIR_FILTER);
+                File[] files2 = file1.listFiles(f -> f.isDirectory() && f.getName().length()==2);
                 for(File file2 : files2) {
-                    File[] files3 = file2.listFiles(FINGERPRINTFILE_FILTER);
+                    File[] files3 = file2.listFiles(f -> f.isFile() && Pattern.compile("[0-9a-f]{28}\\.xml").matcher(f.getName()).matches());
```

Comment:
This will compile the pattern again every iteration of the loop. I think it would be better to move `Pattern.compile("[0-9a-f]{28}\\.xml")` out of the loop so it only happens once. Might be a premature optimization but since it is a doubly-nested loop I feel like it makes sense to just store the compiled pattern in a static field.